File: C/src/binary_search.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

bool binary_search(int ordered_array[], size_t size, int value_to_search)
{
    if (size == 0)
        return false;
    size_t low = 0;
    size_t high = size - 1;

    while (low <= high)
    {
        size_t mid = low + (high - low) / 2;
        int mid_value = ordered_array[mid];

        if (mid_value == value_to_search)
            return true;

        if (value_to_search < mid_value)
            high = mid - 1;
        else if (value_to_search > mid_value)
            low = mid + 1;
    }
    return false;
}

void run_test(int arr[], size_t size, int value_to_search, bool expected_outcome)
{
    static int test_count = 0;
    printf("\n---\n");
    bool res_1 = binary_search(arr, size, value_to_search);
    printf("res_%d: %d\n", test_count++, res_1);
    assert(res_1 == expected_outcome);
}

int main()
{
    int arr_1[] = {1, 2, 3, 4};
    run_test(arr_1, sizeof(arr_1) / sizeof(int), 3, true);

    int arr_2[] = {1, 2, 3, 5, 6};
    run_test(arr_2, sizeof(arr_2) / sizeof(int), 4, false);

    int arr_3[] = {1, 2, 3, 5, 6};
    run_test(arr_3, sizeof(arr_3) / sizeof(int), 4, false);

    int arr_4[] = {4, 4, 4, 4, 4};
    run_test(arr_4, sizeof(arr_4) / sizeof(int), 5, false);

    int arr_5[] = {4, 4, 5, 5, 6, 6};
    run_test(arr_5, sizeof(arr_5) / sizeof(int), 5, true);

    int *arr_6 = malloc(4);
    run_test(arr_6, 0, 5, false);
    free(arr_6);

    int arr_7[] = {4};
    run_test(arr_7, sizeof(arr_7) / sizeof(int), 4, true);

    int arr_8[] = {4, 5};
    run_test(arr_8, sizeof(arr_8) / sizeof(int), 5, true);
}

File: Zig/src/binary-search.zig
const std = @import("std");
const testing = std.testing;
const debug = std.debug;

pub fn binarySearch(arr: []const i32, value: i32) ?usize {
    if (arr.len == 0) {
        return null;
    }
    var low: usize = 0;
    var high: usize = arr.len - 1;
    while (low <= high) {
        const mid: usize = low + (high - low) / 2;
        if (arr[mid] == value) {
            return mid;
        }

        if (value < arr[mid]) {
            if (mid == 0) break;
            high = mid - 1;
        } else if (value > arr[mid]) {
            low = mid + 1;
        }
    }
    return null;
}

test "30 in [0 2 2 30 40] -> 3" {
    const search_arr = [_]i32{ 0, 2, 2, 30, 40 };
    const result = binarySearch(&search_arr, 30).?;
    try testing.expect(result == 3);
}

test "1 in [1] -> 0" {
    const search_arr = [_]i32{1};
    const result = binarySearch(&search_arr, 1).?;
    try testing.expect(result == 0);
}

test "1 in [2, 3, 4] -> null" {
    const search_arr = [_]i32{ 2, 3, 4 };
    const result = binarySearch(&search_arr, 1);
    try testing.expect(result == null);
}

test "5 in [5, 7, 10, 10, 15] -> 0" {
    const search_arr = [_]i32{ 5, 7, 10, 10, 15 };
    const result = binarySearch(&search_arr, 5).?;
    try testing.expect(result == 0);
}

test "50 in [5, 7, 10, 10, 15, 50] -> 5" {
    const search_arr = [_]i32{ 5, 7, 10, 10, 15, 50 };
    const result = binarySearch(&search_arr, 50).?;
    try testing.expect(result == 5);
}
